keps/sig-network/1797-configure-fqdn-as-hostname-for-pods/README.md

```diff
@@ -0,0 +1,762 @@
+(see [sethostname(2)](http://man7.org/linux/man-pages/man2/sethostname.2.html)). Kubernetes attempts to include the 
+Pod name as hostname, unless this limit is reached. When the limit is reached, Kubernetes has a series of mechanisms 
+to deal with the issue. These include, truncating Pod hostname when a “Naked” Pod name is longer than 63 bytes, and 
+having an alternative way of generating Pod names when they are part of a Controller, like a Deployment. The proposed 
+feature might still hit the 63 Bytes limit unless we create or adapt similar remediation techniques. Without any 
+remediation, Kubernetes will fail to create the Pod Sandbox and the pod will remain in “ContainerCreating” (Pending status) 
+forever. The feature proposed here will make this issue occur more frequently, as now the whole FQDN would be limited to 63 
+bytes. Next we illustrate the issue with an example of a potential error message, based on an initial draft of this 
+feature (PR [#91035](https://github.com/kubernetes/kubernetes/pull/91035)):
+
+```bash
+$ kubectl get pod
+NAME                                                  READY   STATUS              RESTARTS   AGE
+longpodnametestsaoitfail23423423432wer-547cc5-st6dd   0/1     ContainerCreating   0          52s
+```
+```bash
+$ kubectl describe pod longpodnametestsaoitfail23423423432wer-547cc5-st6dd
+Name:           longpodnametestsaoitfail23423423432wer-547cc5-st6dd
+Namespace:      foo
+...
+...
+Events:
+  Type     Reason                  Age               From                                Message
+  ----     ------                  ----              ----                                -------
+  Normal   Scheduled               16s               default-scheduler                   Successfully assigned foo/longpodnametestsaoitfail23423423432wer-547cc5-st6dd to host.company.com
+  Warning  FailedCreatePodSandBox  1s (x2 over 16s)  kubelet, host.company.com  Failed create pod sandbox: Failed to set FQDN in hostname, Pod hostname longpodnametestsaoitfail23423423432wer-547cc5-st6dd.p1324234234234.foo.svc.test q.company.com is too long (63 characters is the limit).
+```
+
+This failure mode is not great because it might not be apparent to users that their pods are failing. There are two potential ways 
+to improve UX of the failure-mode:
+* Create an admission plugin that calculates the length of the FQDN of the Pod. The problem of this approach is that it might 
+not cover all scenarios, there are many entry points to generate a pod, i.e., deployments, replicasets, CRD, etc. Another problem 
+is that it breaks Kubernetes abstraction layers as we have to make assumptions from the top layer.
+* Create non-retriable errors for pods. Currently failures like the one generated by this kernel hostname limit retry forever. 
+It would be nice if we can define that an error is fatal, then the pod changes to Failed state.
+
+
+```
+<<[UNRESOLVED Will this work on Windows? ]>>
```

Comment:
I don't think I'm going to be able to get all of the answers I need by today (1.19 KEP enhancement freeze deadline) for if this can work with Windows or not. Windows containers cannot be joined to domain but instead can inherit Active Directory domain identities. Because of this I'm not sure what all the implications would be for setting this in Windows containers.
Because the current proposal is to configure this as part of the POD spec it should be easy to ignore/not set for Windows workloads while this is in alpha.
I'll work on getting more information so we can document expected behaviors for Windows workloads during 1.19.

Reply:
We can approve KEP and then have more discussions and testing during implementation. Maybe this feature does not make sense for Windows today? Like you suggest, we could just ignore it during implementation and revisit later when you expand the scope to run Linux containers from Windows. I think @dchen1107 mentioned that not everything works today on Windows and it should not be a problem if this does not work there today?

Reply:
Yea - we can get this approved and document test the Windows story later in 1.19.